Make command immediately available after scaffolding

haroldangenent opened the pull request:
Fixes #32.

danielbachhuber commented:
@haroldangenent Thanks! Can you make sure the command doc is updated accordingly, and there's a corresponding test for this?

haroldangenent commented on inc/ScaffoldPackageCommand.php:
```diff
@@ -76,7 +76,7 @@ public function package( $args, $assoc_args ) {
 			$package_dir = $assoc_args['dir'];
 		} else {
-			$package_dir = WP_CLI::get_runner()->get_packages_dir_path() . 'vendor/' . $assoc_args['name'];
+			$package_dir = WP_CLI::get_runner()->get_packages_dir_path() . 'src/' . $assoc_args['name'];
```
Of course.

haroldangenent commented:
@danielbachhuber Thanks for the review, sorry about the tests. Ran into some issues running tests locally (all tests actually error, see below). Don't have time to finish this right now, I'll get back to this in a few days.

```
PHP Warning:  require(/tmp/wp-cli-home/.wp-cli/packages/vendor/composer/../wp-cli/foo/command.php): failed to open stream: No such file or directory in /tmp/wp-cli-home/.wp-cli/packages/vendor/composer/autoload_real.php on line 66
```

danielbachhuber commented:
> Thanks for the review, sorry about the tests. Ran into some issues running tests locally (all tests actually error, see below).

Just to confirm, you're running the latest WP-CLI version locally?

haroldangenent commented:
Turns out the tests were actually failing, not having anything to do with the local environment. Sadly, this causes most tests to fail of which I'm not sure how to proceed with.

I found a few different type of failures, most are Composer-related and caused when actually running `WP_CLI::run_command( array( 'package', 'install', $package_dir ) );`:

- [ ] `WP_CLI_PACKAGES_DIR` is not working correctly when a relative path is passed. This affects all tests that contain `WP_CLI_PACKAGES_DIR=packages`. (Composer will error out: `Error: Composer could not find the config file: packages/composer.json`)
- [ ] Conflicts between package requirements when a specific `wp-cli` version is loaded (for example, when running `wp scaffold package wp-cli/foo --require_wp_cli='~0.24.0-alpha'`)
- [ ] PHP / Composer failing to include an autoloaded file, probably created (and removed) by the previous test. This causes all subsequent tests to fail (See [Travis](https://travis-ci.org/wp-cli/scaffold-package-command/jobs/169740614#L276) or my previous comment)
- [x] Some tests containing references to `vendor/` (addressed in https://github.com/wp-cli/scaffold-package-command/pull/63/commits/529b3d719df1ecc3307df06a88f64a4a8a559ffd)

Hope you can give me some guidance, since I'm not sure about the impact of the necessary changes.

**Docs**
Updated the docs in https://github.com/wp-cli/scaffold-package-command/pull/63/commits/419ea02dfec3441d5efc6f45bd72285dd2694b03, but package path did not seem to be documented extensively.

danielbachhuber commented:
> `WP_CLI_PACKAGES_DIR` is not working correctly when a relative path is passed. This affects all tests that contain `WP_CLI_PACKAGES_DIR=packages`

What if we try dropping use of `WP_CLI_PACKAGES_DIR` entirely? I think I only used it originally to have a predictable path for file exists assertions, but we can get the default `PACKAGES_DIR` path in a different manner:

```
When I run `wp package path`
Then save STDOUT as {PACKAGE_PATH}
```

haroldangenent commented:
> What if we try dropping use of WP_CLI_PACKAGES_DIR entirely?

Thanks, that seems like the proper solution. See https://github.com/wp-cli/scaffold-package-command/pull/63/commits/b59dbf4741828eceb7b33b0a1d9de80a90b6fc8a.

**Existing packages**
It seemed that a lot of scenario's were 'hanging' on `File already exists. Overwrite?`. Each scenario of [scaffold-package](https://github.com/wp-cli/scaffold-package-command/blob/master/features/scaffold-package.feature) seems to try and overwrite the package created by the previous scenario.

The easiest solution for now was naming packages for each scenario (https://github.com/wp-cli/scaffold-package-command/pull/63/commits/2bfa9b7364d0661cb102c786c44e5a0a8b0e5801 and https://github.com/wp-cli/scaffold-package-command/pull/63/commits/b5a6ea7acd0a3cdb897d646ce9ff2c567bf7d03c, as done [here as well] (https://github.com/wp-cli/scaffold-package-command/blob/master/features/scaffold-package.feature#L151)), but it feels like a bad solution. The source of the issue is probably that the files are now created in the tests home directory (that persists through multiple scenario's), instead of their own test directory (since `WP_CLI_PACKAGES_DIR` is removed). How do you feel about this?

**Error in `autoload.php`**
[Failing scenario](https://github.com/wp-cli/scaffold-package-command/blob/master/features/scaffold-package-readme.feature#L3-L23).

My assumption: `foo/<package>` is created, but [removed after this scenario](https://github.com/wp-cli/scaffold-package-command/blob/master/features/bootstrap/FeatureContext.php#L113). Composer's autoloader is not updated and tries to require a non-existent file, causing each subsequent scenario to fail.

[Fixed](https://github.com/wp-cli/scaffold-package-command/pull/63/commits/b4d1b64e8fff4d89278dae2fb57deee8838930e9) by not specifying `--dir`. Since `--dir` is not an essential part of the scenario, _I took the liberty of removing it_.


**Conflict between versions**
Only happens in [this scenario](https://github.com/wp-cli/scaffold-package-command/blob/master/features/scaffold-package-readme.feature#L25-L43), but I'm not sure what the origin of the issue is. Hope you can point me in the right direction on this one.

**`--require`**
Right now, most tests use `--require` when actually testing the command, but this shouldn't be necessary anymore since https://github.com/wp-cli/scaffold-package-command/pull/63/commits/d8e827de221dbfb49aabe133d80dc179b6ddd1d0. Would you recommend adjusting the existing tests or creating a new one / several new ones?

danielbachhuber commented:
> Only happens in this scenario, but I'm not sure what the origin of the issue is. Hope you can point me in the right direction on this one.

Try `>=0.24.0-alpha` instead of `~0.24.0-alpha`

> Would you recommend adjusting the existing tests or creating a new one / several new ones?

Let's go ahead and create a new one so we have coverage in both directions.

haroldangenent commented:
@danielbachhuber Summarizing my latest commits: Packages now get uninstalled after each scenario that installs a package. This:

- Handles `--dir` tests properly (making sure Composer doesn't try and autoload non-existent files on the next scenario)
- Makes sure each scenario is independent of the previous scenario's (`wp hello-world` might work because the previous scenario also scaffolded `wp hello-world`)

Tests are now passing, but CircleCI seems to be running out of memory when installing a package. Could be increased with something like this in `circle.yml`:

```
dependencies:
  cache_directories:
    - ~/.composer/cache
  pre:
    # Set the PHP timezone so that Behat does not fail.
    - echo "date.timezone = 'US/Central'" > /opt/circleci/php/$(phpenv global)/etc/conf.d/wp-cli-timezone.ini
    # Disable xdebug, which makes Composer slower.
    - echo ""  > /opt/circleci/php/$(phpenv global)/etc/conf.d/xdebug.ini
    # Increase memory limit.
    - echo "memory_limit = 512M" > /opt/circleci/php/$(phpenv global)/etc/conf.d/wp-cli-memory.ini
```

Running tests takes much longer now than before this PR, since actually installing each package with Composer takes some time.

Thanks for your patience.

danielbachhuber commented:
> Makes sure each scenario is independent of the previous scenario's (`wp hello-world` might work because the previous scenario also scaffolded `wp hello-world`)

Did you actually run into this problem? All of the relevant directories should be blown away between each scenario, so this shouldn't be possible.

> Tests are now passing, but CircleCI seems to be running out of memory when installing a package. 

I think this must be a symptom of another problem.

danielbachhuber commented on features/scaffold-package.feature:
```diff
@@ -121,40 +156,61 @@ Feature: Scaffold WP-CLI commands
+    And the {PACKAGE_PATH}/local/wp-cli/with-tests/features/bootstrap/ProcessRun.php file should exist
 
-    When I run `WP_CLI_PACKAGES_DIR=packages wp --require=packages/vendor/wp-cli/foo/command.php hello-world`
+    When I run `wp --require={PACKAGE_PATH}/local/wp-cli/with-tests/command.php hello-world`
```
This `--require` statement shouldn't be necessary now, because the command should be autoloaded.

haroldangenent replied:
> Let's go ahead and create a new one so we have coverage in both directions.

I have created a new scenario without `--require` and left this one in place as it was. Or would you prefer I adjust the existing tests?

The new test would probably become obsolete in that case.

danielbachhuber commented on features/scaffold-package.feature:
```diff
@@ -3,66 +3,96 @@ Feature: Scaffold WP-CLI commands
+    And the {PACKAGE_PATH}/local/wp-cli/foo/.travis.yml file should not exist
 
-    When I run `WP_CLI_PACKAGES_DIR=packages wp --require=packages/vendor/wp-cli/foo/command.php hello-world`
+    When I run `wp --require={PACKAGE_PATH}/local/wp-cli/foo/command.php hello-world`
```
This `--require` statement shouldn't be necessary now, because the command is autoloaded.

danielbachhuber commented on inc/ScaffoldPackageCommand.php:
```diff
@@ -110,6 +110,8 @@ public function package( $args, $assoc_args ) {
 			WP_CLI::run_command( array( 'scaffold', 'package-readme', $package_dir ), array( 'force' => $force ) );
 		}
+
+		WP_CLI::run_command( array( 'package', 'install', $package_dir ) );
```
Could we potentially use `WP_CLI::launch_self()` here, so installation happens as a part of a new process?

haroldangenent replied:
I've been playing with this and done some reading, but this doesn't seem to work out. From [the docs](https://wp-cli.org/docs/internal-api/wp-cli-launch-self/#notes):

> Note: While this command does persist a limited set of runtime arguments, it does not persist environment variables. Practically speaking, WP-CLI packages won’t be loaded when using WP_CLI::launch_self() because the launched process doesn’t have access to the current process $HOME.

This was exactly the problem in this case as well (`Package_Command->get_composer_json_path()` couldn't locate the existing packages `composer.json` and tries to create one).

haroldangenent commented:
> All of the relevant directories should be blown away between each scenario, so this shouldn't be possible.

Yes, I ran into this issue in https://github.com/wp-cli/scaffold-package-command/pull/63/commits/24777336777189b712ace8e18a43665132533658. This test should fail, but it didn't (because the previous scenario is still in place). See [this scenario in Travis](https://travis-ci.org/wp-cli/scaffold-package-command/jobs/170379220#L597).

All temporary directories are blown away (e.g. `/wp-cli-test-run-<hash>/`), but `/wp-cli-home/` seems to persist through all scenario's.

> I think this must be a symptom of another problem.

CircleCI seems to have [started running out of memory](https://circleci.com/gh/wp-cli/scaffold-package-command/83) in https://github.com/wp-cli/scaffold-package-command/pull/63/commits/d8e827de221dbfb49aabe133d80dc179b6ddd1d0 for `wp scaffold package`. This led me to believe the actual installation causes the memory limit to be hit.

danielbachhuber commented:
> All temporary directories are blown away (e.g. `/wp-cli-test-run-<hash>/`), but `/wp-cli-home/` seems to persist through all scenario's.

Oh :( The test suite in this repo hasn't been updated. This problem will be fixed with https://github.com/wp-cli/scaffold-package-command/pull/64

danielbachhuber commented:
> This problem will be fixed with #64

Specifically https://github.com/wp-cli/scaffold-package-command/pull/64/files#diff-8a50a8f3567ab9a03107fd86eeaf827eR130

haroldangenent commented:
Thanks. I'm not sure about the next steps forward for this PR. Possibly:

> Oh :( The test suite in this repo hasn't been updated. This problem will be fixed with #64

Merge `master` to this branch, remove `package uninstall` for scenario's not using `--dir`?

> CircleCI seems to have [started running out of memory](https://circleci.com/gh/wp-cli/scaffold-package-command/83) in https://github.com/wp-cli/scaffold-package-command/pull/63/commits/d8e827de221dbfb49aabe133d80dc179b6ddd1d0 for `wp scaffold package`.

To be honest, I'm not sure how to proceed on this. Any ideas?

danielbachhuber commented:
@haroldangenent Sorry for dropping the ball on these questions. Still want to see this PR through?

haroldangenent commented:
@danielbachhuber Yes, I think this would be a welcome addition and would be glad to contribute where I can.

danielbachhuber commented:
@haroldangenent Can you try adding the following to `circle.yml`:

```
dependencies:
  pre:
    - echo "memory_limit = 256M" > /opt/circleci/php/$(phpenv global)/etc/conf.d/memory.ini
```

haroldangenent commented:
@danielbachhuber Still had to update test suite from `master`, that's why the tests were failing yesterday (files weren't removed after first test).

danielbachhuber commented:
👍  Thanks for the time you spent on this!
